homeassistant/components/tuya/base.py

```diff
@@ -149,6 +151,93 @@ def available(self) -> bool:
         """Return if the device is available."""
         return self.device.online
 
+    def find_dpcode(
+        self, dpcodes: DPCode | tuple[DPCode, ...] | None, function: bool = False
+    ) -> DPCode | None:
+        """Find a matching DP code available on for this device."""
+        if dpcodes is None:
+            return None
+
+        if not isinstance(dpcodes, tuple):
+            dpcodes = (dpcodes,)
+
+        for dpcode in dpcodes:
+            # If not a function try to use status
+            if not function and dpcode in self.device.status:
+                return dpcode
+
+            # If a function, or (if status not found), try to use function
+            if dpcode in self.device.function:
+                return dpcode
+
+        return None
+
+    def get_dptype(
+        self, dpcode: DPCode | None, function: bool = False
+    ) -> DPType | None:
+        """Find a matching DPCode data type available on for this device."""
+        if dpcode is None:
+            return None
+
+        order = (
+            ["function", "status_range"] if function else ["status_range", "function"]
+        )
+        for key in order:
+            if dpcode in getattr(self.device, key):
+                return DPType(getattr(self.device, key).get(dpcode).type)
+
+        return None
+
+    def get_integer_type(
+        self, dpcodes: DPCode | tuple[DPCode, ...] | None, function: bool = False
+    ) -> IntegerTypeData | None:
+        """Return IntegerTypeData for a found matching DPCode.
+
+        By default we find anything we can use, however we prefer to use status
+        when function is False and prefer function if function is True.
+        """
+        if (dpcode := self.find_dpcode(dpcodes, function)) is None:
+            return None
+
+        order = (
+            ["function", "status_range"] if function else ["status_range", "function"]
+        )
+        for key in order:
+            if (
+                dpcode in getattr(self.device, key)
+                and getattr(self.device, key).get(dpcode).type == "Integer"
+            ):
+                return IntegerTypeData.from_json(
+                    dpcode, getattr(self.device, key)[dpcode].values
+                )
+
+        return None
+
+    def get_enum_type(
+        self, dpcodes: DPCode | tuple[DPCode, ...] | None, function: bool = False
+    ) -> EnumTypeData | None:
+        """Return EnumTypeData for a found matching DPCode.
+
+        By default we find anything we can use, however we prefer to use status
+        when function is False and prefer function if function is True.
+        """
+        if (dpcode := self.find_dpcode(dpcodes, function)) is None:
```

Comment:
Can you make a generic helper that just takes EnumDataType class? All other functionality seems similar for each type?

Reply:
Done, find_dpcode now handles all of these things